### PR TITLE
Lagt til henting og lagring av behandler som kan motta dialogmelding (for aktiv fastlege)

### PR DESCRIPTION
**Pull request by andersrognstad**
Ved kall for å hente behandler hentes aktiv fastlege med partnerId for arbeidstaker.
Det sjekkes om det finnes behandler i databasen med samme partnerId (og om den er knyttet til arbeidstakeren). 
- Dersom behandler finnes vil denne returneres (og knyttes til arbeidstakeren hvis den ikke er det eller ikke er siste behandler knyttet til til arbeidstakeren)
- Dersom behandler ikke finnes vil denne lagres, knyttes til arbeidstakeren og returneres.

### CODE REVIEW

**Review by cyanboy — APPROVED** [review]

File: src/test/kotlin/no/nav/syfo/testhelper/TestDatabase.kt [hunk 1]
```diff
@@ -38,3 +44,37 @@ class TestDatabaseNotResponding : DatabaseInterface {
...
+            connection.createBehandlerDialogmelding(fastlege.toNewBehandler())
+        connection.createBehandlerDialogmeldingArbeidstaker(
+            arbeidstakerPersonIdent,
+            behandlerDialogmelding.first
+        )
+        connection.commit()
+
+        return behandlerDialogmelding.second
+    }
+}
```
Comment (janzh) [inline src/test/kotlin/no/nav/syfo/testhelper/TestDatabase.kt:63]: Er det mulig å bruke "ekte" kode i stedet for denne?

File: src/main/kotlin/no/nav/syfo/behandler/domain/Fastlege.kt [hunk 2]
```diff
@@ -27,21 +29,20 @@ data class Fastlege(
...
+    partnerId = this.partnerId.toString(),
+    herId = this.herId.toString(),
+    parentHerId = this.parentHerId.toString(),
+    hprId = this.helsepersonellregisterId,
+    kontor = this.kontor.navn,
+    adresse = this.kontor.postadresse.adresse,
+    postnummer = this.kontor.postadresse.postnummer,
+    poststed = this.kontor.postadresse.poststed,
+    orgnummer = this.kontor.orgnummer,
+    telefon = this.kontor.telefon
```
Comment (janzh) [inline src/main/kotlin/no/nav/syfo/behandler/domain/Fastlege.kt:47]: Pirk: ikke fjerne trailing komma her.

File: src/test/kotlin/no/nav/syfo/behandler/BehandlerServiceSpek.kt [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            )
+        } returns response
+    }
+
+    describe("BehandlerService") {
+        with(TestApplicationEngine()) {
+            start()
+
+            val fastlegeClientMock = mockk<FastlegeClient>()
+            val partnerinfoClientMock = mockk<PartnerinfoClient>()
```
Comment (janzh) [inline src/test/kotlin/no/nav/syfo/behandler/BehandlerServiceSpek.kt:57]: Hadde vært nice med mock på server-nivå her i stedet for mockk, for best mulig testdekning, men kan jo evt. ta det i en senere PR.

**Review by janzh — APPROVED** [review]

File: src/test/kotlin/no/nav/syfo/behandler/BehandlerServiceSpek.kt (hunk 3 above)
Comment (andersrognstad) [reply src/test/kotlin/no/nav/syfo/behandler/BehandlerServiceSpek.kt:57]: Hmja. Bruker jo servermocks i API-speken, men hadde lyst på en egen service-test for å dekke flere corner-caser der og da virket det greiest med mockk. Men kan undersøke det i ny PR 👍

File: src/test/kotlin/no/nav/syfo/testhelper/TestDatabase.kt (hunk 1 above)
Comment (andersrognstad) [reply src/test/kotlin/no/nav/syfo/testhelper/TestDatabase.kt:63]: Skal vel være mulig å bruke BehandlerService#createBehandlerDialogmelding hvis den ikke er private. 
Litt usikker på om vi skal droppe private kun for testen, men kan vel tenkes at den også kan skal brukes til å opprette andre behandlere enn fastlege på sikt så mulig det får være greit...

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt [hunk 4]
```diff
@@ -33,4 +61,67 @@ class BehandlerService(
...
+
+        if (behandlerDialogmeldingForPersonsFastlege == null) {
+            return createBehandlerDialogmelding(
+                personIdentNumber = personIdentNumber,
+                newBehandler = fastlege.toNewBehandler()
+            )
+        }
+
+        val (arbeidstakerPersonident, pBehandlerDialogmelding) = behandlerDialogmeldingForPersonsFastlege
+        val isNyFastlege = behandlerDialogmeldingForPersonList.indexOf(behandlerDialogmeldingForPersonsFastlege) > 0
```
Comment (geir-waagboe) [inline src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:81]: Hm, skal det ikke være < 0 her? (dvs at behandleren ikke er blant de eksisterende behandlerene for personIdent'en?

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt [hunk 5]
```diff
@@ -33,4 +61,67 @@ class BehandlerService(
...
+            behandlerDialogmeldingForPersonList.find { (_, pBehandlerDialogmelding) -> pBehandlerDialogmelding.partnerId == fastlege.partnerId.toString() }
+
+        if (behandlerDialogmeldingForPersonsFastlege == null) {
+            return createBehandlerDialogmelding(
+                personIdentNumber = personIdentNumber,
+                newBehandler = fastlege.toNewBehandler()
+            )
+        }
+
+        val (arbeidstakerPersonident, pBehandlerDialogmelding) = behandlerDialogmeldingForPersonsFastlege
```
Comment (geir-waagboe) [inline src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:80]: Denne linja bør flyttes ned til etter den neste siden variablene ikke brukes der.

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt [hunk 6]
```diff
@@ -33,4 +61,67 @@ class BehandlerService(
...
+    private fun createOrGetBehandler(fastlege: Fastlege, personIdentNumber: PersonIdentNumber): Behandler {
+        val behandlerDialogmeldingForPersonList = database.getBehandlerDialogmeldingForPerson(
+            personIdentNumber = personIdentNumber,
+        )
+
+        val behandlerDialogmeldingForPersonsFastlege =
+            behandlerDialogmeldingForPersonList.find { (_, pBehandlerDialogmelding) -> pBehandlerDialogmelding.partnerId == fastlege.partnerId.toString() }
+
+        if (behandlerDialogmeldingForPersonsFastlege == null) {
+            return createBehandlerDialogmelding(
```
Comment (geir-waagboe) [inline src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:74]: Må du ikke først sjekke om behandleren finnes (dvs finnes, men ikke blant dem som er knyttet til personIdentNumber)?

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt (hunk 4 above)
Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:81]: Tror det skal bli riktig at vi sjekker om behandleren er ulik siste lagret behandler. For å få historikk på bytte av fastlege (også bytte tilbake til en fastlege man har hatt tidligere). Kanskje tydeligere hvis jeg skriver om til:
`val isBytteAvFastlege = behandlerDialogmeldingForPersonList.first() != behandlerDialogmeldingForPersonsFastlege` ?

**Review by JMLindseth — COMMENTED** [review]
Nå har vi minst 4 objekter som representerer en Behandler (pluss et par for fastlege), jeg klarer ikke å se at det er nødvendig? `NewBehandler` brukes - så vidt jeg kan se - kun som en "mellomlagring" mellom Fastlege og lagring i databasen. Den mappingen kan vel gjøres direkte i queryen?

Deretter gjør man om til `Behandler` som har `id` og `behandlerRef` i tillegg til de samme som `NewBehandler`. BehandlerRef kan lages hvor som helst, men "id" får man ikke før etter lagring i DB, så dette må man jo ha, så sant id faktisk brukes utenfor databasen.

Og så har man `PBehandlerDialogmelding`, den er helt lik `Behandler`, bortsett fra `createdAt`/`updatedAt`, men de brukes ikke i dag, så da er vel heller ikke objektet nødvendig?

Og DTO-objektet som også er helt likt, men mangler id og behandlerRef (men som skal få behandlerRef snart).

Oppsummert: Jeg tror vi kan kvitte oss med mange objekter og mapping og styr, og heller sende Fastlege rett inn i databasespørringer, og så få et Behandler-objekt tilbake som brukes overalt. Og så får vi heller revurdere hvilke objekter som er nødvendige, når vi får inn andre behandlere. 🤷‍♂️

File: src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            updated_at
+            ) VALUES (DEFAULT, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) RETURNING id
+    """
+
+const val queryGetBehandlerDialogmeldingForArbeidstaker =
+    """
+        SELECT BEHANDLER_DIALOGMELDING_ARBEIDSTAKER.arbeidstaker_personident, BEHANDLER_DIALOGMELDING.*
+        FROM
+        BEHANDLER_DIALOGMELDING
+        LEFT JOIN BEHANDLER_DIALOGMELDING_ARBEIDSTAKER
```
Comment (geir-waagboe) [inline src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt:108]: left join  sammen med betingelsen ellers betyr så vidt jeg forstår at du får alle radene i behandler_dialogmelding i resultsettet. Hadde det ikke vært bra å ha partner_id som en betingelse i tillegg?

**Review by JMLindseth — APPROVED** [review]

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt (hunk 6 above)
Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:74]: Det gjøres vel nå siden spørringen gjør left join og tar med seg også behandlere som ikke er knyttet til personident. Men som du skriver under er det nok bedre å ta partnerId inn i spørringen - skal se på det.

File: src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt (hunk 7 above)
Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt:108]: Ja, riktig, jeg ser det nå. Hadde partner_id som en betingelse til å begynne med, men fjernet den (husker ikke helt hvorfor 😅 ). Bør nok endres. Kanskje er det ryddigere å ha én spørring som bare sjekker mot behandler_dialogmelding på partnerId og en ny som sjekker om en behandler_dialogmelding er knyttet til personident 🤔

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt (hunk 4 above)
Comment (geir-waagboe) [reply src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:81]: Ah, ok så listen er ordnet slik at den nyeste har index=0? Ja, det blir kanskje enklere å forstå med forslaget ditt

**Comment by andersrognstad** [comment]
> Nå har vi minst 4 objekter som representerer en Behandler (pluss et par for fastlege), jeg klarer ikke å se at det er nødvendig? `NewBehandler` brukes - så vidt jeg kan se - kun som en "mellomlagring" mellom Fastlege og lagring i databasen. Den mappingen kan vel gjøres direkte i queryen?
> 
> Deretter gjør man om til `Behandler` som har `id` og `behandlerRef` i tillegg til de samme som `NewBehandler`. BehandlerRef kan lages hvor som helst, men "id" får man ikke før etter lagring i DB, så dette må man jo ha, så sant id faktisk brukes utenfor databasen.
> 
> Og så har man `PBehandlerDialogmelding`, den er helt lik `Behandler`, bortsett fra `createdAt`/`updatedAt`, men de brukes ikke i dag, så da er vel heller ikke objektet nødvendig?
> 
> Og DTO-objektet som også er helt likt, men mangler id og behandlerRef (men som skal få behandlerRef snart).
> 
> Oppsummert: Jeg tror vi kan kvitte oss med mange objekter og mapping og styr, og heller sende Fastlege rett inn i databasespørringer, og så få et Behandler-objekt tilbake som brukes overalt. Og så får vi heller revurdere hvilke objekter som er nødvendige, når vi får inn andre behandlere. 🤷‍♂️

Ja, tanken var å legge til rette for å lagre andre behandlere enn fastlege. Men vi kan sikkert forenkle litt nå og heller se an, i hvert fall kvitte oss med NewBehandler. Synes det er greit å ha egne objekter for database, domene og api. Dvs beholde PBehandlerDialogmelding, Behandler og DTO-en (kommer til å fjerne diverse ider og legge til behandlerRef her i ny PR).

File: src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt (hunk 7 above)
Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt:108]: Med inner join vil man vel ikke få ut behandlere som finnes, men som er knyttet til en annen personident (dvs de hvor vi bare skal opprette kobling)? Sånn sett er det vel greiest å dele opp i to spørringer.

**Comment by JMLindseth** [comment]
> > Nå har vi minst 4 objekter som representerer en Behandler (pluss et par for fastlege), jeg klarer ikke å se at det er nødvendig? `NewBehandler` brukes - så vidt jeg kan se - kun som en "mellomlagring" mellom Fastlege og lagring i databasen. Den mappingen kan vel gjøres direkte i queryen?
> > Deretter gjør man om til `Behandler` som har `id` og `behandlerRef` i tillegg til de samme som `NewBehandler`. BehandlerRef kan lages hvor som helst, men "id" får man ikke før etter lagring i DB, så dette må man jo ha, så sant id faktisk brukes utenfor databasen.
> > Og så har man `PBehandlerDialogmelding`, den er helt lik `Behandler`, bortsett fra `createdAt`/`updatedAt`, men de brukes ikke i dag, så da er vel heller ikke objektet nødvendig?
> > Og DTO-objektet som også er helt likt, men mangler id og behandlerRef (men som skal få behandlerRef snart).
> > Oppsummert: Jeg tror vi kan kvitte oss med mange objekter og mapping og styr, og heller sende Fastlege rett inn i databasespørringer, og så få et Behandler-objekt tilbake som brukes overalt. Og så får vi heller revurdere hvilke objekter som er nødvendige, når vi får inn andre behandlere. 🤷‍♂️
> 
> Ja, tanken var å legge til rette for å lagre andre behandlere enn fastlege. Men vi kan sikkert forenkle litt nå og heller se an, i hvert fall kvitte oss med NewBehandler. Synes det er greit å ha egne objekter for database, domene og api. Dvs beholde PBehandlerDialogmelding, Behandler og DTO-en (kommer til å fjerne diverse ider og legge til behandlerRef her i ny PR).

I teorien er jeg enig i at det er greit å skille mellom db, intern, og api, men det er jo bare mer styr når det ikke er noe forskjell på objektene, og de kun brukes som "mellomlagring". Poenget med å ha ulike objekter, er jo at man ikke nødvendigvis trenger samme informasjon internt som det man eksponerer for brukere av APIet, men når vi trenger samme informasjon overalt, er det ikke noe verdi i å skille på de.

**Comment by janzh** [comment]
> Nå har vi minst 4 objekter som representerer en Behandler (pluss et par for fastlege), jeg klarer ikke å se at det er nødvendig? `NewBehandler` brukes - så vidt jeg kan se - kun som en "mellomlagring" mellom Fastlege og lagring i databasen. Den mappingen kan vel gjøres direkte i queryen?
> 
> Deretter gjør man om til `Behandler` som har `id` og `behandlerRef` i tillegg til de samme som `NewBehandler`. BehandlerRef kan lages hvor som helst, men "id" får man ikke før etter lagring i DB, så dette må man jo ha, så sant id faktisk brukes utenfor databasen.
> 
> Og så har man `PBehandlerDialogmelding`, den er helt lik `Behandler`, bortsett fra `createdAt`/`updatedAt`, men de brukes ikke i dag, så da er vel heller ikke objektet nødvendig?
> 
> Og DTO-objektet som også er helt likt, men mangler id og behandlerRef (men som skal få behandlerRef snart).
> 
> Oppsummert: Jeg tror vi kan kvitte oss med mange objekter og mapping og styr, og heller sende Fastlege rett inn i databasespørringer, og så få et Behandler-objekt tilbake som brukes overalt. Og så får vi heller revurdere hvilke objekter som er nødvendige, når vi får inn andre behandlere. 🤷‍♂️

Nå har jeg ikke sett nøye på disse objektene i denne appen, men det er i hvert fall fordelaktig å ikke gjenbruke objekter på tvers av lagene i domenet, dvs. at man ikke gjenbruke objekter på tvers av database-laget, app-laget eller DTO-lag. Dette gjelder selv om noen av objektene kan ha samme felter i dag. Ved å måtte eksplisitt mappe ting når man skal ut  av eller inn i applikasjons-laget gjør at man unngår å blø ut av for mye informasjon til f.eks en konsument av et REST-api eller Kafka-topic ved en feiltakelse.

**Comment by andersrognstad** [comment]
> > > Nå har vi minst 4 objekter som representerer en Behandler (pluss et par for fastlege), jeg klarer ikke å se at det er nødvendig? `NewBehandler` brukes - så vidt jeg kan se - kun som en "mellomlagring" mellom Fastlege og lagring i databasen. Den mappingen kan vel gjøres direkte i queryen?
> > > Deretter gjør man om til `Behandler` som har `id` og `behandlerRef` i tillegg til de samme som `NewBehandler`. BehandlerRef kan lages hvor som helst, men "id" får man ikke før etter lagring i DB, så dette må man jo ha, så sant id faktisk brukes utenfor databasen.
> > > Og så har man `PBehandlerDialogmelding`, den er helt lik `Behandler`, bortsett fra `createdAt`/`updatedAt`, men de brukes ikke i dag, så da er vel heller ikke objektet nødvendig?
> > > Og DTO-objektet som også er helt likt, men mangler id og behandlerRef (men som skal få behandlerRef snart).
> > > Oppsummert: Jeg tror vi kan kvitte oss med mange objekter og mapping og styr, og heller sende Fastlege rett inn i databasespørringer, og så få et Behandler-objekt tilbake som brukes overalt. Og så får vi heller revurdere hvilke objekter som er nødvendige, når vi får inn andre behandlere. 🤷‍♂️
> > 
> > 
> > Ja, tanken var å legge til rette for å lagre andre behandlere enn fastlege. Men vi kan sikkert forenkle litt nå og heller se an, i hvert fall kvitte oss med NewBehandler. Synes det er greit å ha egne objekter for database, domene og api. Dvs beholde PBehandlerDialogmelding, Behandler og DTO-en (kommer til å fjerne diverse ider og legge til behandlerRef her i ny PR).
> 
> I teorien er jeg enig i at det er greit å skille mellom db, intern, og api, men det er jo bare mer styr når det ikke er noe forskjell på objektene, og de kun brukes som "mellomlagring". Poenget med å ha ulike objekter, er jo at man ikke nødvendigvis trenger samme informasjon internt som det man eksponerer for brukere av APIet, men når vi trenger samme informasjon overalt, er det ikke noe verdi i å skille på de.

Jeg synes likevel det er fint å ha eksplisitt skille på objektene i de ulike lagene selv om de "tilfeldigvis" er like nå. Det gjør at man ved endringer blir nødt til å ta stilling til hva som hører hjemme de ulike lagene og evt skrive litt mapping-kode. (Og mapping-kode er jo enkelt 😁 )

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt (hunk 4 above)
Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt]: Refaktorert dette nå

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt (hunk 6 above)
Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:70]: Refaktorert til to spørringer og endret service tilsvarende nå.

File: src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt (hunk 7 above)
Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/database/BehandlerDialogmeldingQuery.kt]: Fikset!

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt [hunk 8]
```diff
@@ -1,16 +1,46 @@
...
+        personIdentNumber: PersonIdentNumber,
+        token: String,
+        callId: String,
+    ): List<Behandler> {
+        val aktivFastlege = getFastlegeMedPartnerinfo(
+            personIdentNumber = personIdentNumber,
+            token = token,
+            callId = callId,
+        )
+        if (aktivFastlege != null) {
```
Comment (geir-waagboe) [inline src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:31]: Litt smak og behag, men jeg hadde foretrukket return utenfor if'en her.

Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:31]: Kan endre til return if her og i getFastlegeMedPartnerinfo

File: src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt [hunk 9]
```diff
@@ -33,4 +63,59 @@ class BehandlerService(
...
+            return createBehandlerDialogmelding(
+                personIdentNumber = personIdentNumber,
+                fastlege = fastlege
+            )
+        }
+
+        val pBehandlereForArbeidstakerList =
+            database.getBehandlerDialogmeldingForArbeidstaker(personIdentNumber = personIdentNumber)
+
+        val isBytteAvFastlege = pBehandlereForArbeidstakerList.firstOrNull() != pBehandlerForFastlege
```
Comment (geir-waagboe) [inline src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:79]: Synes det er litt fishy å basere seg på equals metoden i data-klassen her. Kanskje bedre å sammenlikne pBehandlerForFastlege.id?

Comment (andersrognstad) [reply src/main/kotlin/no/nav/syfo/behandler/BehandlerService.kt:79]: Enig. Gjelder vel både i sjekkene != og contains her. Strengt tatt kunne man vel bare returnert id-er og eventuelt behandlerRef fra database.getBehandlerDialogmeldingForArbeidstaker 🤔

**Review by geir-waagboe — APPROVED** [review]
